io/src/lib.rs

```diff
@@ -230,6 +230,57 @@ impl<T: AsRef<[u8]>> BufRead for Cursor<T> {
     }
 }
 
+/// A [`Read`]er which keeps an internal buffer to avoid hitting the underlying stream directly for
+/// every read, implementing [`BufRead`].
+pub struct BufReader<'a, R: Read> {
+    inner: &'a mut R,
+    buf: [u8; 4096],
+    pos: usize,
+    limit: usize,
+}
+
+impl<'a, R: Read> BufReader<'a, R> {
+    /// Creates a [`BufReader`] which will read from the given `inner`.
+    pub fn new(inner: &'a mut R) -> Self {
+        BufReader {
+            inner,
+            buf: [0; 4096],
+            pos: 0,
+            limit: 0,
+        }
+    }
+}
+
+impl<'a, R: Read> Read for BufReader<'a, R> {
+    #[inline]
+    fn read(&mut self, output: &mut [u8]) -> Result<usize> {
+        let input = self.fill_buf()?;
+        let count = cmp::min(input.len(), output.len());
+        output[..count].copy_from_slice(&input[..count]);
+        self.consume(count);
+        Ok(count)
+    }
+}
+
+impl<'a, R: Read> BufReader<'a, R> {
+    #[inline]
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        if self.pos < self.limit {
+            Ok(&self.buf[self.pos..self.limit])
+        } else {
+            let count = self.inner.read(&mut self.buf[..])?;
+            self.pos = 0;
+            self.limit = cmp::min(count, self.buf.len());
```

Comment:
If `count > self.buf.len()` we should panic instead since it's clearly a bug. (But maybe `debug_assert` is enough.)

Reply:
I don't think silently coercing the value is helpful. It's most likely harmful since it just produces garbage data. Kinda like UB. Now I also realize that we can either omit the check or turn it into `assert!` because indexing below will panic anyway. I think a nice panic message saying the `Read` implementation is broken would be nicer. (The bound check below will be eliminated by the optimizer in that case.)

Reply:
If we replace `min` with `debug_assert` then `count > self.buf.len()` implies `self.limit > self.buf.len()` which implies panic on `self.buf[..self.limit]`.

Reply:
I'm proposing to remove it. It's just trying to pretend a clearly broken code is not really broken hoping for the best. I don't think we should do that in financial applications.

Reply:
I think we should use a real `assert`. If we benchmark to find that this is hurting performance we can put the `debug_assert`/`min` combo back.